### PR TITLE
LibJS+Everywhere: Propagate OOM errors from `Cell::initialize` and `Heap::allocate`

### PR DESCRIPTION
**Pull request by trflynn89**
One big hurdle we have in the `DeprecatedString` -> `String` conversion is `PropertyKey` (and `StringOrSymbol`). `PropertyKey`s are created ad-nauseam in `initialize` overrides, so we will need to propagate errors from creating the `String`s that will form the `PropertyKey`s.

So this makes `initialize` (and it's caller in `allocate`) handle completions. And at the end, propagates one OOM error we were masking in `StringObject::initialize` as seen here:

Screenshot from 2023-01-28 13-55-32

### CODE REVIEW

**Review by linusg — COMMENTED** [review]
Awesome! One naming suggestion, and shouldn't `allocate()` and `initialize()` exclusively use `MUST_OR_THROW_OOM`? (which I do realize is probably a PITA to change after the fact :grimacing:)

File: Userland/Libraries/LibJS/Runtime/Completion.h [hunk 1]
```diff
@@ -332,6 +332,12 @@ class [[nodiscard]] ThrowCompletionOr {
     [[nodiscard]] ValueType release_value() { return m_value.release_value(); }
     Completion release_error() { return m_throw_completion.release_value(); }
 
+    ValueType release_allocated_value_but_fixme_should_propagate_errors()
```
[the remark this answers is not in the document]
Comment (trflynn89) [reply Userland/Libraries/LibJS/Runtime/Completion.h:335]: I was on the fence about both this name and using `MUST_OR_THROW_OOM`. I figured for this method, I wanted to really dissuade anyone from using it for anything *other* than failures to `allocate`, after all the work we put into getting rid of `TRY_OR_DISCARD`.

For `TRY`, I wasn't sure if we wanted to throw the hammer that you're only allowed to throw in `initialize` if you OOM?

[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/Completion.h:335]: > I figured for this method, I wanted to really dissuade anyone from using it for anything _other_ than failures to `allocate`, after all the work we put into getting rid of `TRY_OR_DISCARD`.

Fair enough, I already forgot about that macro :sweat_smile:

> For `TRY`, I wasn't sure if we wanted to throw the hammer that you're only allowed to throw in `initialize` if you OOM?

Is there any other legitimate use case? I can't think of any, especially given it's a non-standard thing...

[the remark this answers is not in the document]
Comment (trflynn89) [reply Userland/Libraries/LibJS/Runtime/Completion.h:335]: > Is there any other legitimate use case? I can't think of any, especially given it's a non-standard thing...

:man_shrugging: They're all `MUST_OR_THROW_OOM` now

**Review by linusg — APPROVED** [review]
Thanks for the `Object` -> `Base` changes as well :)

**Pull request merged**